src/liberator_mixin/hal/core.clj

```diff
@@ -45,30 +45,22 @@
 
     [halboy.json :as haljson]
 
-    [hype.core :as hype]
-
     [liberator-mixin.logging.core :as log]
     [jason.convenience :as jason-conv])
   (:import
-    [halboy.resource Resource]
-    [java.util UUID]))
-
-(defn- random-uuid []
-  (str (UUID/randomUUID)))
+    [halboy.resource Resource]))
 
 (def hal-media-type
   "The HAL media type string."
   "application/hal+json")
 
 (extend-protocol r/Representation
   Resource
-  (as-response [data {:keys [request routes] :as context}]
-    (r/as-response
-      (-> data
-        (hal/add-link :discovery
-          (hype/absolute-url-for request routes :discovery))
-        (haljson/resource->map))
-      context)))
+  (as-response [data {:keys [discovery-url-fn] :as context}]
+    (let [resource (if discovery-url-fn
+                     (hal/add-link data :discovery (discovery-url-fn))
```

Comment:
It looks like the self ref wasn't tested before, but can you please add some.

Reply:
Have added a couple of tests, for with and and without the discovery-url-fn. 👍